x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.xpack.esql.core.expression.Alias;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.Count;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.CountDistinct;
+import org.elasticsearch.xpack.esql.plan.logical.Aggregate;
+import org.elasticsearch.xpack.esql.plan.logical.Eval;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.Project;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Replaces an aggregation function having a false/null filter with an EVAL node.
+ * <pre>
+ *     ... | STATS x = someAgg(y) WHERE FALSE {BY z} | ...
+ *     =>
+ *     ... | STATS x = someAgg(y) {BY z} > | EVAL x = NULL | KEEP x{, z} | ...
+ * </pre>
+ */
+public class ReplaceStatsFilteredAggWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
+    @Override
+    protected LogicalPlan rule(Aggregate aggregate) {
+        List<NamedExpression> newAggs = new ArrayList<>(aggregate.aggregates().size());
+        List<Alias> newEvals = new ArrayList<>(aggregate.aggregates().size());
+        List<NamedExpression> newProjections = new ArrayList<>(aggregate.aggregates().size());
+        for (var ne : aggregate.aggregates()) {
+            if (ne instanceof Alias alias
+                && alias.child() instanceof AggregateFunction aggFunction
+                && aggFunction.hasFilter()
+                && aggFunction.filter() instanceof Literal literal
+                && literal.fold().equals(false)) {
+
+                Object value = aggFunction instanceof Count || aggFunction instanceof CountDistinct ? 0L : null;
+                Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
+                newEvals.add(newAlias);
+                newProjections.add(newAlias.toAttribute());
```

Comment:
I got different results between the branch and main for a query like below, I wonder if we should keep the aggregation here? 🤔

main
```
+ Q2_FalseFilter
{
  "query": "from sample_data | stats c = count(*) where false, s = sum(event.duration) where false"
}
'
       c       |       s
---------------+---------------
0              |null
+ Q3_NullFilter
{
  "query": "from sample_data | stats c = count(*) where null, s = sum(event.duration) where null"
}
'
       c       |       s
---------------+---------------
0              |null
+ Q4_TrueFilter
{
  "query": "from sample_data | stats c = count(*) where true, s = sum(event.duration) where true"
}
'
       c       |       s
---------------+---------------
7              |23231327
```


branch
```
+ Q2_FalseFilter
{
  "query": "from sample_data | stats c = count(*) where false, s = sum(event.duration) where false"
}
'
       c       |       s
---------------+---------------
0              |null
0              |null
0              |null
0              |null
0              |null
0              |null
0              |null
+ Q3_NullFilter
{
  "query": "from sample_data | stats c = count(*) where null, s = sum(event.duration) where null"
}
'
       c       |       s       
---------------+---------------
0              |null           
0              |null        
0              |null          
0              |null
0              |null
0              |null           
0              |null           
+ Q4_TrueFilter
{ 
  "query": "from sample_data | stats c = count(*) where true, s = sum(event.duration) where true"
}              
'
       c       |       s
---------------+---------------
7              |23231327       
```

Reply:
I think the aggregate gets replaced with a projection that returns the constant for each entry:
`FROM sample__data | stats c = count(*) where null, sum = sum() where null  | eval c = 0, sum = null`
becomes
`FROM sample_data | eval c = 0, sum = null`

Reply:
Thanks, indeed, the case where the Aggregation was entirely pruned wasn't handled correctly. It's been fixed  now.